pkg/server/plugin/keymanager/gcpkms/gcpkms.go

```diff
@@ -0,0 +1,1040 @@
+			Labels:  cryptoKeyLabels,
+			Purpose: kmspb.CryptoKey_ASYMMETRIC_SIGN,
+			VersionTemplate: &kmspb.CryptoKeyVersionTemplate{
+				Algorithm: algorithm,
+			},
+		},
+		CryptoKeyId: cryptoKeyID,
+		Parent:      config.KeyRing,
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create CryptoKey: %v", err)
+	}
+
+	log := p.log.With(cryptoKeyNameTag, cryptoKey.Name)
+	log.Debug("CryptoKey created", algorithmTag, algorithm)
+
+	if err := p.setIamPolicy(ctx, cryptoKey.Name); err != nil {
+		log.Debug("Failed to set IAM policy")
+		return nil, status.Errorf(codes.Internal, "failed to set IAM policy: %v", err)
+	}
+
+	cryptoKeyVersionName := cryptoKey.Name + "/cryptoKeyVersions/1"
+	log.Debug("CryptoKeyVersion version added", cryptoKeyVersionNameTag, cryptoKeyVersionName)
+
+	pubKey, err := getPublicKeyFromCryptoKeyVersion(ctx, p.kmsClient, cryptoKeyVersionName)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get public key: %v", err)
+	}
+	newKeyEntry := keyEntry{
+		cryptoKey:            cryptoKey,
+		cryptoKeyVersionName: cryptoKeyVersionName,
+		publicKey: &keymanagerv1.PublicKey{
+			Id:          spireKeyID,
+			Type:        keyType,
+			PkixData:    pubKey,
+			Fingerprint: makeFingerprint(pubKey),
+		},
+	}
+
+	p.entries[spireKeyID] = newKeyEntry
+	return &newKeyEntry, nil
+}
+
+// disposeCryptoKeys looks for active CryptoKeys that haven't been updated
+// during the maxStaleDuration time window. Those keys are then enqueued for
+// destruction.
+func (p *Plugin) disposeCryptoKeys(ctx context.Context) error {
+	p.log.Debug("Looking for CryptoKeys to dispose")
+
+	config, err := p.getConfig()
+	if err != nil {
+		return err
+	}
+
+	disposeCryptoKeysFilter, err := p.getDisposeCryptoKeysFilter()
+	if err != nil {
+		return err
+	}
+	itCryptoKeys := p.kmsClient.ListCryptoKeys(ctx, &kmspb.ListCryptoKeysRequest{
+		Parent: config.KeyRing,
+		Filter: disposeCryptoKeysFilter,
+	})
+
+	for {
+		cryptoKey, err := itCryptoKeys.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			p.log.Error("Failure listing CryptoKeys to dispose", reasonTag, err)
+			return err
+		}
+
+		itCryptoKeyVersions := p.kmsClient.ListCryptoKeyVersions(ctx, &kmspb.ListCryptoKeyVersionsRequest{
+			Parent: cryptoKey.Name,
+			Filter: "state = " + kmspb.CryptoKeyVersion_ENABLED.String(),
+		})
+
+		// If the CryptoKey doesn't have any enabled CryptoKeyVersion, mark it
+		// as inactive so it's not returned future calls.
+		cryptoKeyVersion, err := itCryptoKeyVersions.Next()
+		if errors.Is(err, iterator.Done) {
+			p.setInactive(ctx, cryptoKey)
+			continue
+		}
+
+		for {
+			if err != nil {
+				p.log.Error("Failure listing CryptoKeyVersios", reasonTag, err)
+				return err
+			}
+
+			if err := p.enqueueDestruction(cryptoKeyVersion.Name); err != nil {
+				p.log.With(cryptoKeyNameTag, cryptoKey.Name).Error("Failed to enqueue CryptoKeyVersion for destruction", reasonTag, err)
+			}
```

Comment:
That's a good point. Calling setInactive after detecting that there are no more CryptoKeyVersions at this point assumed that all versions could be enqueued for destruction, but that may not always be the case if there was an error. Since we already call setInactive if the returned CryptoKey doesn't have enabled CryptoKeyVersions, a safer approach would be to just leave that and remove this one.